program/src/processor.rs

```diff
@@ -407,66 +406,48 @@ fn process_harvest_rewards(program_id: &Pubkey, accounts: &[AccountInfo]) -> Pro
         bytemuck::try_from_bytes_mut::<HolderRewards>(&mut holder_rewards_data)
             .map_err(|_| ProgramError::InvalidAccountData)?;
 
-    // Update the holder rewards state with the new "last seen" total rewards.
-    let last_seen_total_rewards = std::mem::replace(
-        &mut holder_rewards_state.last_seen_total_rewards,
-        current_total_rewards,
-    );
-
-    // Calculate unharvested rewards for the token account.
-    // Since the holder rewards account may already have unharvested rewards,
-    // calculate the share of rewards that have not been seen by the holder
-    // rewards account.
-    let rewards_to_harvest = {
-        // Rewards accrued by the system that haven't been seen by the holder.
-        let unseen_total_rewards = current_total_rewards
-            .checked_sub(last_seen_total_rewards)
-            .ok_or(ProgramError::ArithmeticOverflow)?;
-
-        // The holder's share of the unseen rewards.
-        let unseen_unharvested_rewards =
-            calculate_reward_share(token_supply, token_account_balance, unseen_total_rewards)?;
-
-        // The total unharvested rewards the holder is eligible to claim.
-        let total_eligible_rewards = holder_rewards_state
-            .unharvested_rewards
-            .checked_add(unseen_unharvested_rewards)
-            .ok_or(ProgramError::ArithmeticOverflow)?;
-
-        // Temporarily update unharvested rewards to reflect total eligible
-        // rewards. This will be re-adjusted post-harvest.
-        holder_rewards_state.unharvested_rewards = total_eligible_rewards;
-
-        // Make sure the pool can't be overdrawn by harvesting.
+    // Determine the amount the holder can harvest.
+    //
+    // This is done by subtracting the `last_accumulated_rewards_per_token`
+    // rate from the pool's current rate, then multiplying by the token account
+    // balance.
+    //
+    // If the pool doesn't have enough lamports to cover the rewards, only
+    // harvest the available lamports. This should never happen, but the check
+    // is a failsafe.
+    let pool_excess_lamports = {
         let rent = <Rent as Sysvar>::get()?;
         let rent_exempt_lamports = rent.minimum_balance(std::mem::size_of::<HolderRewardsPool>());
-
-        // If the pool doesn't have enough lamports to cover the rewards, only
-        // harvest the available lamports.
-        total_eligible_rewards.min(
-            holder_rewards_pool_info
-                .lamports()
-                .checked_sub(rent_exempt_lamports)
-                .ok_or(ProgramError::ArithmeticOverflow)?,
-        )
+        holder_rewards_pool_info
+            .lamports()
+            .saturating_sub(rent_exempt_lamports)
```

Comment:
nit: Is there a reason to change this from checked to saturating?

Reply:
I figured since we do the `min()` later and then if it's zero do nothing, it was fine to saturate into zero rather than error.
Also, wouldn't it be not possible for this account to be holding valid state and to have less than rent-exempt at this point? We aren't debiting it in the processor before this point, and should this proceed with non-zero we'd get the "results in account below rent-exempt" error.

Reply:
Yeah it's fine either way, entirely pedantic. We can leave it as is. I was more wondering why it changed, but I think that was more because the diff showed it oddly.